File: ios/Runner/MediaPipe/HandAnalyzerBridge.swift
import Foundation
import MediaPipeTasksVision
import UIKit

/// Wraps MediaPipe Tasks [HandLandmarker] in IMAGE running mode.
///
/// The `.task` model file must be bundled into the Runner target's resources:
///   1. Drag `hand_landmarker.task` into the Runner folder in Xcode.
///   2. Ensure "Copy items if needed" and the Runner target checkbox are ticked.
final class HandAnalyzerBridge {
    private let landmarker: HandLandmarker

    init() throws {
        let modelPath = try Self.requireModelPath(name: "hand_landmarker", ext: "task")
        let options = HandLandmarkerOptions()
        options.baseOptions.modelAssetPath = modelPath
        options.runningMode = .image
        options.numHands = 2
        options.minHandDetectionConfidence = 0.5
        options.minHandPresenceConfidence = 0.5
        options.minTrackingConfidence = 0.5
        landmarker = try HandLandmarker(options: options)
    }

    func detect(frame: FrameArgs) throws -> [[String: Any]] {
        guard let uiImage = FrameDecoder.decodeUprightImage(frame: frame) else {
            return []
        }
        let mpImage = try MPImage(uiImage: uiImage)
        let result = try landmarker.detect(image: mpImage)

        let width = Double(uiImage.size.width)
        let height = Double(uiImage.size.height)

        var hands: [[String: Any]] = []
        for (i, landmarkList) in result.landmarks.enumerated() {
            let handCategory = i < result.handedness.count ? result.handedness[i].first : nil
            let label = handCategory?.categoryName ?? "Unknown"
            let score = Double(handCategory?.score ?? 0)
            let points = landmarkList.map { lm -> [Double] in
                [Double(lm.x) * width, Double(lm.y) * height]
            }
            hands.append([
                "handedness": label,
                "confidence": score,
                "landmarks": points,
            ])
        }
        return hands
    }

    private static func requireModelPath(name: String, ext: String) throws -> String {
        guard let path = Bundle.main.path(forResource: name, ofType: ext) else {
            throw NSError(
                domain: "MediaPipePlugin",
                code: 404,
                userInfo: [NSLocalizedDescriptionKey: "Missing bundled model: \(name).\(ext)"],
            )
        }
        return path
    }
}

File: android/app/src/main/kotlin/com/example/face_liveness_demo/mediapipe/HandAnalyzerBridge.kt
package com.example.face_liveness_demo.mediapipe

import android.content.Context
import android.graphics.Bitmap
import com.google.mediapipe.framework.image.BitmapImageBuilder
import com.google.mediapipe.tasks.core.BaseOptions
import com.google.mediapipe.tasks.vision.core.RunningMode
import com.google.mediapipe.tasks.vision.handlandmarker.HandLandmarker
import com.google.mediapipe.tasks.vision.handlandmarker.HandLandmarkerResult

/**
 * Wraps MediaPipe Tasks [HandLandmarker] in IMAGE running mode.
 *
 * Model file must be bundled at `android/app/src/main/assets/hand_landmarker.task`.
 */
class HandAnalyzerBridge(context: Context) {
    private val landmarker: HandLandmarker

    init {
        val options = HandLandmarker.HandLandmarkerOptions.builder()
            .setBaseOptions(
                BaseOptions.builder()
                    .setModelAssetPath(MODEL_ASSET_PATH)
                    .build(),
            )
            .setRunningMode(RunningMode.IMAGE)
            .setNumHands(2)
            .setMinHandDetectionConfidence(0.5f)
            .setMinHandPresenceConfidence(0.5f)
            .setMinTrackingConfidence(0.5f)
            .build()
        landmarker = HandLandmarker.createFromOptions(context, options)
    }

    /**
     * Detect hands in [frame] and return a serializable payload:
     *   [ { handedness, confidence, landmarks: [[x_px, y_px], ...] }, ... ]
     *
     * Landmark coordinates are returned in pixels of the rotation-applied
     * upright frame — so the Flutter side does not need to do any rotation math.
     */
    fun detect(frame: FrameArgs): List<Map<String, Any>> {
        val bitmap = FrameDecoder.decodeUprightBitmap(frame) ?: return emptyList()
        val mpImage = BitmapImageBuilder(bitmap).build()
        val result: HandLandmarkerResult = landmarker.detect(mpImage)

        val width = bitmap.width
        val height = bitmap.height
        val hands = mutableListOf<Map<String, Any>>()

        for (i in 0 until result.landmarks().size) {
            val landmarkList = result.landmarks()[i]
            val handednessList = if (i < result.handedness().size) result.handedness()[i] else emptyList()
            val categoryConfidence = handednessList.firstOrNull()?.score()?.toDouble() ?: 0.0
            val categoryLabel = handednessList.firstOrNull()?.categoryName() ?: "Unknown"

            val points = landmarkList.map { lm ->
                listOf(
                    (lm.x() * width).toDouble(),
                    (lm.y() * height).toDouble(),
                )
            }
            hands += mapOf(
                "handedness" to categoryLabel,
                "confidence" to categoryConfidence,
                "landmarks" to points,
            )
        }
        return hands
    }

    fun close() {
        try {
            landmarker.close()
        } catch (_: Throwable) {
        }
    }

    companion object {
        private const val MODEL_ASSET_PATH = "hand_landmarker.task"
    }
}
